plugins/auth/irods_auth_plugin-native/libirods_auth_plugin-native.cpp

```diff
@@ -0,0 +1,467 @@
+
+            // build a buffer for the challenge hash
+            // Buffer structure:
+            //  [64+1] challenge string (generated by server and sent back here)
+            //  [50+1] obfuscated user password
+            char md5_buf[CHALLENGE_LEN + MAX_PASSWORD_LEN + 2]{};
+            strncpy(md5_buf, request_result.c_str(), CHALLENGE_LEN);
+
+            // Save a representation of some of the challenge string for use as a session signature
+            setSessionSignatureClientside(md5_buf);
+
+            // determine if a password challenge is needed, are we anonymous or not?
+            bool need_password = false;
+            if (req.at("user_name") == ANONYMOUS_USER) {
+                md5_buf[CHALLENGE_LEN + 1] = '\0';
+            }
+            else {
+                need_password = obfGetPw(md5_buf + CHALLENGE_LEN);
+            }
+
+            // prompt for a password if necessary
+            if (need_password) {
+                struct termios tty;
+                memset( &tty, 0, sizeof( tty ) );
+                tcgetattr( STDIN_FILENO, &tty );
+                tcflag_t oldflag = tty.c_lflag;
+                tty.c_lflag &= ~ECHO;
+                int error = tcsetattr( STDIN_FILENO, TCSANOW, &tty );
+                int errsv = errno;
+
+                if (error) {
+                    std::cout << "WARNING: Error "
+                              << errsv
+                              << " disabling echo mode. Password will be displayed in plaintext.\n";
+                }
+                printf( "Enter your current iRODS password:" );
+                std::string password{};
+                getline(std::cin, password);
+                strncpy(md5_buf + CHALLENGE_LEN, password.c_str(), MAX_PASSWORD_LEN);
+                tty.c_lflag = oldflag;
+                if (tcsetattr(STDIN_FILENO, TCSANOW, &tty)) {
+                    printf( "Error reinstating echo mode." );
+                }
+            }
+
+            // create a md5 hash of the challenge
+            MD5_CTX context;
+            MD5_Init( &context );
+            MD5_Update(&context, (unsigned char*)md5_buf, CHALLENGE_LEN + MAX_PASSWORD_LEN);
+
+            char digest[RESPONSE_LEN + 2]{};
+            MD5_Final((unsigned char*)digest, &context);
+
+            // make sure 'string' doesn't end early - scrub out any errant terminating chars
+            // by incrementing their value by one
+            for (int i = 0; i < RESPONSE_LEN; ++i) {
+                if (digest[i] == '\0') {
+                    digest[i]++;
+                }
+            }
+
+            unsigned char out[RESPONSE_LEN*2];
+            unsigned long out_len{RESPONSE_LEN*2};
+            auto err = base64_encode((unsigned char*)digest, RESPONSE_LEN, out, &out_len);
+            if(err < 0) {
+                THROW(err, "base64 encoding of digest failed.");
+            }
+
+            resp["digest"] = std::string{(char*)out};
```

Comment:
this assumes `out` is null-terminated.

if the `out_len` represents the length of the string, we should pass it to the std::string constructor.

Reply:
Will do